library/md.c

```diff
@@ -928,69 +901,77 @@ const int *mbedtls_md_list(void)
-{
-    if (NULL == md_name) {
-        return NULL;
-    }
+typedef struct {
+    const char *md_name;
+    mbedtls_md_type_t md_type;
+} md_name_entry;
 
-    /* Get the appropriate digest information */
+static const md_name_entry md_names[] = {
 #if defined(MBEDTLS_MD_CAN_MD5)
-    if (!strcmp("MD5", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_MD5);
-    }
+    { "MD5", MBEDTLS_MD_MD5 },
 #endif
 #if defined(MBEDTLS_MD_CAN_RIPEMD160)
-    if (!strcmp("RIPEMD160", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_RIPEMD160);
-    }
+    { "RIPEMD160", MBEDTLS_MD_RIPEMD160 },
 #endif
 #if defined(MBEDTLS_MD_CAN_SHA1)
-    if (!strcmp("SHA1", md_name) || !strcmp("SHA", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_SHA1);
-    }
+    { "SHA1", MBEDTLS_MD_SHA1 },
+    { "SHA", MBEDTLS_MD_SHA1 }, // compatibility fallback
 #endif
 #if defined(MBEDTLS_MD_CAN_SHA224)
-    if (!strcmp("SHA224", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_SHA224);
-    }
+    { "SHA224", MBEDTLS_MD_SHA224 },
 #endif
 #if defined(MBEDTLS_MD_CAN_SHA256)
-    if (!strcmp("SHA256", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_SHA256);
-    }
+    { "SHA256", MBEDTLS_MD_SHA256 },
 #endif
 #if defined(MBEDTLS_MD_CAN_SHA384)
-    if (!strcmp("SHA384", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_SHA384);
-    }
+    { "SHA384", MBEDTLS_MD_SHA384 },
 #endif
 #if defined(MBEDTLS_MD_CAN_SHA512)
-    if (!strcmp("SHA512", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_SHA512);
-    }
+    { "SHA512", MBEDTLS_MD_SHA512 },
 #endif
 #if defined(MBEDTLS_MD_CAN_SHA3_224)
-    if (!strcmp("SHA3-224", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_SHA3_224);
-    }
+    { "SHA3-224", MBEDTLS_MD_SHA3_224 },
 #endif
 #if defined(MBEDTLS_MD_CAN_SHA3_256)
-    if (!strcmp("SHA3-256", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_SHA3_256);
-    }
+    { "SHA3-256", MBEDTLS_MD_SHA3_256 },
 #endif
 #if defined(MBEDTLS_MD_CAN_SHA3_384)
-    if (!strcmp("SHA3-384", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_SHA3_384);
-    }
+    { "SHA3-384", MBEDTLS_MD_SHA3_384 },
 #endif
 #if defined(MBEDTLS_MD_CAN_SHA3_512)
-    if (!strcmp("SHA3-512", md_name)) {
-        return mbedtls_md_info_from_type(MBEDTLS_MD_SHA3_512);
-    }
+    { "SHA3-512", MBEDTLS_MD_SHA3_512 },
 #endif
-    return NULL;
+    { NULL, MBEDTLS_MD_NONE },
+};
+
+const mbedtls_md_info_t *mbedtls_md_info_from_string(const char *md_name)
+{
+    if (NULL == md_name) {
+        return NULL;
+    }
+
+    const md_name_entry *entry = md_names;
+    while (entry->md_name != NULL &&
+           strcmp(entry->md_name, md_name) != 0) {
+        ++entry;
+    }
+
+    return mbedtls_md_info_from_type(entry->md_type);
+}
+
+const char *mbedtls_md_get_name(const mbedtls_md_info_t *md_info)
```

Comment:
Maybe we need to update the description in https://github.com/Mbed-TLS/mbedtls/blob/a36ef6b41080c96ad8ea6f862541113f9ccaf1a0/include/mbedtls/md.h#L467-L468

Reply:
I don't think the public documentation needs to reflect how the function works internally, as long as the effects are the same.

Reply:
Yes, but it is actually wrong now. It should really say something along the lines of "This function returns the name of the message digest for the message-digest information structure given"

Reply:
Ok, right. I'll make the change.